### PR TITLE
Add retries to zigbee device discovery

### PR DESCRIPTION
**Pull request by karlkar**
Solves issue #92

### CODE REVIEW

**Comment by syssi** [comment]
Are you aware of these issues? #26 #45 

IMO your PR is just a workaround for a single case. It should be handled properly later on, too. Do I miss something?

**Comment by Danielhiversen** [comment]
Closing this

**Comment by andriej** [comment]
@Danielhiversen @syssi I have 3 gateways which are going 'unavailable' after some problem occurs.
Anyway - restarting HA doesn't get them working in HA back.
In traffic debug I see that discovery is being 'done':
`Feb 16 12:41:11 hacek hass[26307]: 2019-02-16 12:41:11 INFO (SyncWorker_14) [xiaomi_gateway] Xiaomi Gateway 34ce008cxxxx found at IP 192.168.x.xx`
although next MCAST packets get to HA:

all entities and gateway is in yellow in lovelace/unavailable and component error.
Restarts don't help, the only thing that helps (?!?) is turning OFF and ON (power restart) of gateways itself.

How to get it fixed? What could be the issue? Getting all sensors unavailable at least once/twice a day is a no-go, including water-flood sensors....

Please guys, let me help get data and get this solved. Just tell me what's needed.

**Comment by syssi** [comment]
Could you update your PR?

File: xiaomi_gateway/__init__.py [hunk 1]
```diff
@@ -261,7 +265,12 @@ def _discover_devices(self):
 
         for sid in sids:
             cmd = '{"cmd":"read","sid":"' + sid + '"}'
-            resp = self._send_cmd(cmd, "read_ack") if int(self.proto[0:1]) == 1 else self._send_cmd(cmd, "read_rsp")
+            trycount = 0
+            while trycount < self._discovery_retries:
```
Comment (syssi) [inline xiaomi_gateway/__init__.py]: Please use `for retry in range(self._discovery_retries):`

File: xiaomi_gateway/__init__.py [hunk 2]
```diff
@@ -46,10 +46,10 @@ def discover_gateways(self):
             host = gateway.get('host')
             port = gateway.get('port')
             sid = gateway.get('sid')
+            discovery_retries = gateway.get('discovery_retries', 1)
```
Comment (syssi) [inline xiaomi_gateway/__init__.py]: Should we use a better default here? 4 retries f.e.

Comment (andriej) [reply xiaomi_gateway/__init__.py]: 3 minimum in my opinion.
It shouldn't make initialization much longer anyway?

Comment (karlkar) [reply xiaomi_gateway/__init__.py]: That would make much more sense, however I wanted it to work as before by default.

Comment (syssi) [reply xiaomi_gateway/__init__.py]: The retry loop breaks early if the zigbee device is discovered properly. The startup/discovery will take some more time for the bad case only.

**Comment by syssi** [comment]
@Danielhiversen I would like to merge this PR. Do you agree?

**Comment by andriej** [comment]
How about merge and pushing it to 0.89 HA? :-)

**Pull request merged**